packages/satcheljs/lib/createUndo.ts

```diff
@@ -0,0 +1,185 @@
+                    undo: () => { modifiedObject[event.name] = event.oldValue; }
+                };
+            }
+            break;
+        case 'splice':
+            undoStep = {
+                verify: () => {
+                    for (let i = 0; i < event.addedCount; i++) {
+                        if (modifiedObject[event.index + i] !== event.added[i]) {
+                            return false;
+                        }
+                    }
+                    return true;
+                },
+                objectName: modifiedObject.$mobx.name,
+                propertyName: event.index,
+                undo: () => {
+                    // First, remove the added items.
+                    // Then, add items back one at a time, because passing an array in to 'splice' will insert the array as a single item
+                    modifiedObject.splice(event.index, event.addedCount);
+                    for (let i = 0; i < event.removedCount; i++) {
+                        modifiedObject.splice(event.index + i, 0, event.removed[i]);
+                    }
+                }
+            };
+            break;
+        case 'add':
+            if (typeof modifiedObject.get !== 'undefined') {
+                // add (map)
+                undoStep = {
+                    verify: () => modifiedObject.get(event.name) === event.newValue,
+                    objectName: modifiedObject.$mobx.name,
+                    propertyName: event.name,
+                    undo: () => { modifiedObject.delete(event.name); }
+                }
+            } else {
+                // add (object)
+                undoStep = {
+                    verify: () => modifiedObject[event.name] === event.newValue,
+                    objectName: modifiedObject.$mobx.name,
+                    propertyName: event.name,
+                    undo: () => { delete modifiedObject[event.name]; }
+                }
+            }
+            break;
+        case 'delete':
+            undoStep = {
+                verify: () => !modifiedObject.has(event.name),
+                    objectName: modifiedObject.$mobx.name,
+                    propertyName: event.name,
+                    undo: () => { modifiedObject.set(event.name, event.oldValue); }
+            }
+            break;
+        default:
+            // Nothing worth tracking
+            return;
+    }
+
+    undoWindows.forEach(window => window.steps.push(undoStep));
+}
+
+export interface UndoResult<T> {
+    actionReturnValue?: T;
+    (): void;
+}
+
+export type CreateUndoReturnValue<T> = (action: () => T | void) => UndoResult<T>;
+
+function trackUndo<T>(
+    actionName: string,
+    action: () => T,
+    undoVerifiesChanges: boolean) : UndoResult<T> {
+
+        initializeSpy();
+        undoWindows.push({steps: []});
+
+        try {
+            let returnValue: T = action();
+
+            let window : UndoWindow = undoWindows[undoWindows.length - 1];
```

Comment:
Technically this is okay - but using the keyword window as a variable is probably not the best idea